### ISSUE TITLE
Warning: Kaledio is not available on this system. Julia will be unable to produce any plots. (aarch64)

### ISSUE THREAD

**Issue by truedichotomy** [sglyon/PlotlyBase.jl#40]
I'm trying to generate some jpg figures using savefig on aarch64 linux, but it complains that Kaledio (Kaleido?) is not available on this system. Any work around?

**Comment by PetrKryslUCSD** [comment]
I have been also getting this error for a while now. Here is the error message coming from `PlotlyBase`:
```
    Building PlotlyJS → `C:\Users\PetrKrysl\.julia\scratchspaces\44cfe95a-1eb2-52ea-b672-e2afdf69b78f\c7a8796665c4d87b6be1c258d507b1fa8424212c\build.log`                                                                                                                       
  Progress [========================================>]  2/2                                                                             
2 dependencies successfully precompiled in 12 seconds (38 already precompiled)                                                          
                                                                                                                                        
julia> using PlotlyJS                                                                                                                   
┌ Warning: Kaleido is not available on this system. Julia will be unable to produce any plots.                                          
└ @ PlotlyBase C:\Users\PetrKrysl\.julia\packages\PlotlyBase\1axKx\src\kaleido.jl:58                                                    
                                                                                                                                        
(@v1.6) pkg> st                                                                                                                         
Status `C:\Users\PetrKrysl\.julia\environments\v1.6\Project.toml`                                                                       
  [f0f68f2c] PlotlyJS v0.14.0 `https://github.com/JuliaPlots/PlotlyJS.jl.git#master`                                                    
                                                                                                                                        
(@v1.6) pkg> 
```

**Comment by PetrKryslUCSD** [comment]
Was able to locate the artifact:
![image](https://user-images.githubusercontent.com/9855832/106517598-8fc2b280-648d-11eb-890b-302f1402a897.png)

Alas, I have no idea whether that is right or not.

**Comment by PetrKryslUCSD** [comment]
It seems to me the artifact is in the right place. The problem seems to be that access is denied to `bin\kaleido.exe`.
```
C:\Users\PetrKrysl\.julia\artifacts\c959551ad0b7f5c1c33b9cb31baa203bcad3a1af\kaleido>.\bin\kaleido.exe
Access is denied.

C:\Users\PetrKrysl\.julia\artifacts\c959551ad0b7f5c1c33b9cb31baa203bcad3a1af\kaleido>
```
@truedichotomy Could you check whether or not you can execute the file?

**Comment by giordano** [comment]
In order to execute an executable, it needs to have execution permission.  The file `kaleido/bin/kaleido.exe` in https://github.com/sglyon/PlotlyBase.jl/releases/download/v0.3.1/kaleido_win.tar.gz has permissions `444`

**Comment by PetrKryslUCSD** [comment]
I should have checked the properties of the file. I just listed it in bash, and there the permissions seemed to be okay:
![image](https://user-images.githubusercontent.com/9855832/106639945-585e0f80-653a-11eb-9c7e-5cc1cd16f9bf.png)
(Any idea why that is?)

**Comment by giordano** [comment]
What version of Julia do you use?

**Comment by PetrKryslUCSD** [comment]
1.6

**Comment by giordano** [comment]
Not sure what's going on, but the file in the tarballs has definitely the wrong permissions.  We've had similar issues in https://github.com/JuliaLang/julia/issues/38993, the solution is to fix the content of the tarball (which is wrong anyway)

**Comment by PetrKryslUCSD** [comment]
At least on Windows, the artifact is not created with correct permissions:
![image](https://user-images.githubusercontent.com/9855832/106642878-c6f09c80-653d-11eb-9dfa-ad610a53c06b.png)

**Comment by truedichotomy** [comment]
I'm running Julia 1.5.3 on a Raspberry Pi 4 running latest 64-bit RaspberryPi OS.

**Comment by PetrKryslUCSD** [comment]
And on your machine: does the artifact work? Can you run the executable?

**Comment by PetrKryslUCSD** [comment]
@sglyon it seems to me that the permissions for the executable are set on lines [57](https://github.com/sglyon/PlotlyBase.jl/blob/d9b65b0d4999982203c63f8713cf3d7268719b83/create_artifact_script.jl#L57) and [61](https://github.com/sglyon/PlotlyBase.jl/blob/d9b65b0d4999982203c63f8713cf3d7268719b83/create_artifact_script.jl#L61), but only for Linux and the Mac. I believe that we could use the [`chmod`](https://docs.julialang.org/en/v1/base/file/#Base.Filesystem.chmod) Julia function for the Windows target (`chmod(ex1_path, 555)`, ...).

**Comment by PetrKryslUCSD** [comment]
I noticed that one @Markhng  added the script to installed artifact, there was a mention of the supported platforms being Linux and  the Mac. Would be possible to extend the support to Windows please?

**Comment by Markhng** [comment]
Could you please run `versioninfo()` on your device? @PetrKryslUCSD

**Comment by PetrKryslUCSD** [comment]
```
Julia Version 1.6.0-beta1.0
Commit b84990e1ac (2021-01-08 12:42 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: Intel(R) Core(TM) i7-8705G CPU @ 3.10GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-11.0.0 (ORCJIT, skylake)
```

**Comment by Markhng** [comment]
`PlotlyBase` passed all tests (`]test PlotlyBase`) on my Windows x64 device (Julia 1.5). The script #39 I update is to make kaleido support `Linux(:aarch64)` and `Windows(:i686)` (also update the version of kaleido).

There are some ways you can try to solve this problem in my mind:
1. remove pacakge by `]rm PlotlyBase` (or `Plotly`), and `]gc --all`.  Then,  `]add PlotlyBase` in admin cmd;
2. if step 1 not works, after step 1, extract `https://github.com/plotly/Kaleido/releases/download/v0.1.0/kaleido_win_x64.zip` manually in the same artifact folder with same folder name as the older one, (which is `~/.julia/artifacts/c959551ad0b7f5c1c33b9cb31baa203bcad3a1af/kaleido`);
3. try julia 1.5.

**Comment by giordano** [comment]
I _suspect_ (but not 100% sure) that Julia v1.5 used to fix broken permissions of executables, and Julia v1.6 doesn't do that anymore.  In any case, the permissions of the `kaleido` executable in the tarball  are wrong (`444`), fixing them is the most straightforward thing to do.

**Comment by PetrKryslUCSD** [comment]
@Markhng My point was that the permissions were set for the executables for the platform Linux/Mac, but not for the Windows. The executable permissions need to be set for the regular user, not just for the administrator.
![image](https://user-images.githubusercontent.com/9855832/107068590-ac186500-6795-11eb-9258-42bfb653da40.png)
![image](https://user-images.githubusercontent.com/9855832/107068642-b9cdea80-6795-11eb-9445-873ebc2c190a.png)

**Comment by Markhng** [comment]
Maybe you can set the executable permission manually as for now?

**Comment by PetrKryslUCSD** [comment]
I can certainly do that. However, I would have to do that on all systems that I use. And, ALL users on Windows who are not running Julia as administrators will have to do the same. Clearly fixing it at the source would be preferable. Is that a lot of work?

**Comment by Markhng** [comment]
It is a problem, and all the artifact files should also be uploaded to release after executing the modified script. @sglyon

**Comment by PetrKryslUCSD** [comment]
I for one would certainly appreciate it. It is the only thing which prevents me and my collaborators from using 1.6.

**Comment by PetrKryslUCSD** [comment]
BTW the execute permissions needs to be set for EVERYONE (not just the administrator) for both kaleido.exe and kaleido.cmd.

**Comment by sglyon** [comment]
Thanks everyone. I am working on getting a fix out soon

**Comment by deltaeecs** [comment]
> `PlotlyBase` passed all tests (`]test PlotlyBase`) on my Windows x64 device (Julia 1.5). The script #39 I update is to make kaleido support `Linux(:aarch64)` and `Windows(:i686)` (also update the version of kaleido).
> 
> There are some ways you can try to solve this problem in my mind:
> 
> 1. remove pacakge by `]rm PlotlyBase` (or `Plotly`), and `]gc --all`.  Then,  `]add PlotlyBase` in admin cmd;
> 2. if step 1 not works, after step 1, extract `https://github.com/plotly/Kaleido/releases/download/v0.1.0/kaleido_win_x64.zip` manually in the same artifact folder with same folder name as the older one, (which is `~/.julia/artifacts/c959551ad0b7f5c1c33b9cb31baa203bcad3a1af/kaleido`);
> 3. try julia 1.5.

I tried the 2nd strep and covered the origin Kaleido(version 0.0.3) with the latest one and it worked!

**Comment by giordano** [comment]
@truedichotomy @PetrKryslUCSD can you please test
```julia
]add https://github.com/giordano/PlotlyBase.jl.git#mg/kaleido-jll
```
?  This should solve issues  for Windows and aarch64 Linux.  Ref: #42.

**Comment by sglyon** [comment]
Should be fixed now on latest release (v0.5.1)

**Comment by VivaldoMendes** [comment]
Hi, @sglyon and @giordano the link above

`]add https://github.com/giordano/PlotlyBase.jl.git#mg/kaleido-jll `

is not working. I tried to install PlotlyBase on my environment and got the message:

```
 (@v1.6) pkg> add https://github.com/giordano/PlotlyBase.jl.git#mg/kaleido-jll
     Cloning git-repo `https://github.com/giordano/PlotlyBase.jl.git`
    Updating git-repo `https://github.com/giordano/PlotlyBase.jl.git`
ERROR: Did not find rev mg/kaleido-jll in repository
```

I am on a **Windows 10 machine** with `Julia v1.6.1`, no  `PlotlyBase.jl` installed (for reasons above), and with the nbextensions in the right place: `webio-jupyter-notebook enabled  - Validating: ok `. The following Julia packages installed and build:

```
PlotlyJs v.0.14.1
Blink v0.12.5
WebIO v0.8.15
```
When using the **Jupyter-notebook** to produce a simple plot with PlotlyJS, I also get the now famous warning

```
Warning: Kaledio is not available on this system. Julia will be unable to produce any plots.
@ PlotlyBase C:\Users\mende\.julia\packages\PlotlyBase\NxSlF\src\kaleido.jl:58
```
The funny thing is that despite the warning in Jupyter-Notebook, in VSCode I have been able to get the plots coming out without any fuss. 

1. **Not in Atom**: when the code is run, it runs smooth with no errors, but no plot to be seen, neither if I use the savefig() instead
2. **Not in Jupyter-Notebook** (the famous Warning comes up),
3.  **Yes in VSCode** . In case it may help, my VSCode details of are as follows:

```
Version: 1.55.2 (user setup)
Commit: 3c4e3df9e89829dce27b7b5c24508306b151f30d
Date: 2021-04-13T09:35:57.887Z
Electron: 11.3.0
Chrome: 87.0.4280.141
Node.js: 12.18.3
V8: 8.7.220.31-electron.0
OS: Windows_NT x64 10.0.19042
``` 

See MWE below

```
using PlotlyJS
x = LinRange(0,2*pi,32)
X = [x x x]
Y = [cos.(x) cos.(x .+ pi/4) cos.(x .+ pi/2)]
layout = Layout(;title="Plotting with PlotlyJS", titlefont_size=18, width=600, height=400)
p1 = plot(X,Y, mode="markers+lines", line_width=0.5, layout)
```
with the correspondent output:

![aaa_PlotlyJS](https://user-images.githubusercontent.com/48516539/116786301-bd6fc580-aa95-11eb-88f7-5942e753985a.png)

**Just a final remark**. On my Julia environment **v1.4.2**, I can run PlotlyJS in Jupyter notebooks without any problems (I have done it), and possibly in Atom as well (but I have not tried it).

PlotlyJS is such a wonderful package that is a pity to have this recurrent problem with the rendering of plots. The prospects of using PlotlyJS with Pluto look amazing.

**Comment by giordano** [comment]
I deleted the branch in my fork as that was a temporary test.  You should really be using the latest version of the official packages, what's the problem with it?

**Comment by VivaldoMendes** [comment]
The problem is that I am not capable of using PloylyJS in Jupyter notebooks: the plots simply do not pop up.

For teaching activities, the interactive functionalities of PlotlyJS are amazing, but if we cannot use it together with Jupyter notebooks, then it is like having a rocket ... but no fuel. I know that I can use notebooks in VSCode, but it is not the same thing as using Jupyter. For teaching, the Jupyter or Pluto notebooks are hardly replaced by other means (at least by now).

I tried also to install the "official" packages as you mention. I do have the most recent version of PlotlyJS (v.014.1) installed and build on my Windows 10 machine, but unfortunately **it did not install** one of its dependencies: PlotlyBase.jl, which I think is crucial to render the plots. I tried `up PlotlyJS` many times to see if the problem would be corrected, but with no success. Then I tried to directly install the PlotlyBase.jl from the official site and this is what I got: 

```
(@v1.6) pkg> add https://github.com/sglyon/PlotlyBase.jl
    Updating git-repo `https://github.com/sglyon/PlotlyBase.jl`
   Resolving package versions...
ERROR: Unsatisfiable requirements detected for package ORCA [47be7bcc]:
 ORCA [47be7bcc] log:
 ├─possible versions are: 0.1.0-0.5.0 or uninstalled
 ├─restricted to versions * by an explicit requirement, leaving only versions 0.1.0-0.5.0
 └─restricted by compatibility requirements with PlotlyBase [a03496cd] to versions: 
uninstalled — no versions left
   └─PlotlyBase [a03496cd] log:
     ├─possible versions are: 0.5.3 or uninstalled
     └─PlotlyBase [a03496cd] is fixed to version 0.5.3

(@v1.6) pkg> st PlotlyBase
  No Matches in `C:\Users\migue\.julia\environments\v1.6\Project.toml`
```
So, I am able to produce plots with PlotlyJS in VSCode, but not in Jupyter-notebook, where I most need them. It looks that on the side of the Jupyter notebook everything seems to be properly installed. What do I need to do in order to stop receiving the message:
```
Warning: Kaleido is not available on this system. Julia will be unable to save images of any plots.
```

**Comment by VivaldoMendes** [comment]
@sglyon @giordano Thank you, guys. PlotlyJS is working again in Julia 1.6.1.

The two problems I had (in different computers) when using PlotlyJS are solved (Windows 10 machine, Julia 1.6.1). They were:
1.  The package PlotlyBase.jl was missing from my environment, and I could not add it manually, which would lead to the error:
```
Warning: Kaledio is not available on this system. Julia will be unable to produce any plots.
@ PlotlyBase C:\Users\mende\.julia\packages\PlotlyBase\NxSlF\src\kaleido.jl:58
```
2.  In another computer (where the package PlotlyBase.jl was available), the error above did not appear. Instead, the code would run without any error. Still, the plots were not produced/displayed.

**In case someone has a similar problem** (and apparently there are quite a few), this is the way it worked for me on different computers:
1. `]up PlotlyJS`
2. Check whether the package PlotlyBase.jl is available in your environment. If not, add it manually by doing:
`]add https://github.com/sglyon/PlotlyBase.jl`
To successfully install this package, I had to `rm ORCA` (ORCA v.0.5.0) because there was a conflict between PlotlyBase and ORCA.
3. I added (following the suggestion here https://discourse.julialang.org/t/plotlyjs-does-not-display-plot-and-errors-on-pkg-test-plotlyjs/57684)

     `]add https://github.com/jonas-kr/PlotlyJS.jl`

4. `up`

And, finally, after a long struggle, voilà, PlotlyJS is back with its great functionalities. The packages' versions are: 

```
PlotlyJs v.0.14.1
PlotlyBase v0.5.3
Blink v0.12.5
WebIO v0.8.15
```

**Comment by davideamato** [comment]
I'm having the same issue with Julia 1.7 on MacOS 12.0.1. I get the warning
![Screenshot 2021-12-10 at 18 01 02](https://user-images.githubusercontent.com/50680451/145620419-fd3b9cf9-1b40-47b8-843c-072d25b607ea.png)
from Julia REPL, Jupyter notebooks, and VSCode.

It seems that the artifact is not present on my system as I can't find kaleido by searching in `~/.julia/artifacts`.

**Comment by giordano** [comment]
When you report issues with macOS, please specify also the architecture (x86_64 vs aarch64)

**Comment by davideamato** [comment]
> When you report issues with macOS, please specify also the architecture (x86_64 vs aarch64)

Sorry about that, I've edited the original post.

**Comment by shayandavoodii** [comment]
I confronted this issue with this:
![image](https://user-images.githubusercontent.com/52105833/150655155-cf20a49d-e6c4-43cc-ac25-2a22c599289a.png)

**Comment by maucejo** [comment]
Hi all,

Sorry to reopen this issue, but I have the same warning as those reported by @PetrKryslUCSD on Windows 11. More explicitly, I receive the following warning:
```
julia> using PlotlyJS
julia> ┌ Warning: Kaleido is not available on this system. Julia will be unable to save images of any plots.
└ @ PlotlyJS C:\Users\maucejo\.julia\packages\PlotlyJS\Jj38U\src\kaleido.jl:65
┌ Warning: ErrorException("Could not start Kaleido process")
└ @ PlotlyJS C:\Users\maucejo\.julia\packages\PlotlyJS\Jj38U\src\kaleido.jl:66
``` 

Here is the ouput of `versioninfo()`
```
Julia Version 1.8.2
Commit 36034abf26 (2022-09-29 15:21 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: 8 × Intel(R) Core(TM) i7-10610U CPU @ 1.80GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-13.0.1 (ORCJIT, skylake)
  Threads: 1 on 8 virtual cores
```

The output of `] st`
```
Status `C:\Users\maucejo\.julia\environments\v1.8\Project.toml`
  [6e4b80f9] BenchmarkTools v1.3.2
  [13f3f980] CairoMakie v0.9.2
  [a93c6f00] DataFrames v1.4.2
  [e9467ef8] GLMakie v0.7.2
  [7073ff75] IJulia v1.23.3
  [b964fa9f] LaTeXStrings v1.3.0
  [7031d0ef] LazyGrids v0.4.0
  [7ed4a6bd] LinearSolve v1.27.1
  [14b8a8f1] PkgTemplates v0.7.29
  [a03496cd] PlotlyBase v0.8.19
  [f0f68f2c] PlotlyJS v0.18.10
  [91a5bcdd] Plots v1.35.8
 [c3e4b0f8] Pluto v0.19.9
  [33c8b6b6] ProgressLogging v0.1.4
  [295af30f] Revise v3.4.0
  [37e2e46d] LinearAlgebra
  [2f01184e] SparseArrays
```

To try to solve this issue, I tried to:

* Remove PlotlyJS (and clean Julia with `gc --all`) and reinstall PlotlyJS
* Replace the content of the artifact, by the one I download on the github repo of Kaleido

Unfortunately, none of the previous actions solved the problem. I am out of solution. Does someone find a solution ?

Thank you for your help.

**Comment by giordano** [comment]
> Sorry to reopen this issue

Please open a new issue.  And since the warning is coming from https://github.com/JuliaPlots/PlotlyJS.jl/blob/02feab87e848ab89c51b40e47b036e916ec419fa/src/kaleido.jl#L65 you likely want to open in it [`PlotlyJS.jl`](https://github.com/JuliaPlots/PlotlyJS.jl) rather than here.
